src/ast/visitor.h

```diff
@@ -0,0 +1,369 @@
+  }
+  R visit(AssignVarStatement &assignment)
+  {
+    visitModify(&assignment.var);
+    visitModify(&assignment.expr);
+    return default_value();
+  }
+  R visit(AssignConfigVarStatement &assignment)
+  {
+    return visitModify(&assignment.expr);
+  }
+  R visit(VarDeclStatement &decl)
+  {
+    return visitModify(&decl.var);
+  }
+  R visit(If &if_node)
+  {
+    visitModify(&if_node.cond);
+    visitModify(&if_node.if_block);
+    visitModify(&if_node.else_block);
+    return default_value();
+  }
+  R visit(Jump &jump)
+  {
+    return visitModify(&jump.return_value);
+  }
+  R visit(Unroll &unroll)
+  {
+    visitModify(&unroll.expr);
+    visitModify(&unroll.block);
+    return default_value();
+  }
+  R visit(While &while_block)
+  {
+    visitModify(&while_block.cond);
+    visitModify(&while_block.block);
+    return default_value();
+  }
+  R visit(For &for_loop)
+  {
+    visitModify(&for_loop.decl);
+    visitModify(&for_loop.expr);
+    visitImpl(for_loop.stmts);
+    return default_value();
+  }
+  R visit(Predicate &pred)
+  {
+    return visitModify(&pred.expr);
+  }
+  R visit(Probe &probe)
+  {
+    visitImpl(probe.attach_points);
+    visitModify(&probe.pred);
+    visitModify(&probe.block);
+    return default_value();
+  }
+  R visit(Config &config)
+  {
+    visitImpl(config.stmts);
+    return default_value();
+  }
+  R visit(Block &block)
+  {
+    visitImpl(block.stmts);
+    return default_value();
+  }
+  R visit(Subprog &subprog)
+  {
+    visitImpl(subprog.args);
+    visitImpl(subprog.stmts);
+    return default_value();
+  }
+  R visit(Program &program)
+  {
+    visitImpl(program.functions);
+    visitImpl(program.probes);
+    visitModify(&program.config);
+    return default_value();
+  }
+
+  // Temporarily allow visits to expression and statement references. This
+  // does not permit the modification of the underlying value, but does allow
+  // the existing passes to continue to work (which do not modify anything, so
+  // this is not a problem for the time being).
+  template <typename T>
+  R visit(T &t)
+  {
+    auto ptr = &t;
+    if constexpr (!std::is_void_v<R>) {
+      auto rval = visitModify(&ptr);
+      assert(ptr == &t); // Should not be modified.
+      return rval;
+    } else {
+      visitModify(&ptr);
+      assert(ptr == &t); // See above.
+    }
+  }
+
+  // Automatically unpacked and dispatch all variant and vector types into the
+  // suitable visitor method.
+  //
+  // In order to automatically replace a variant, e.g. change from type A to
+  // type B, it is necessary to provide a replace method that accepts that
+  // variant type directly. This could still dispatch via the standard visit
+  // function, which could e.g. return the replacement pointer, but this would
+  // be a single specialized pass for this case.
+  template <typename... Ts>
+  R visit(std::variant<Ts *...> var)
+  {
+    return std::visit([this](auto &value) -> R { return visitModify(&value); },
```

Comment:
Naming question: should this be `visitReplace` ?

Reply:
I've made this `visitAndReplace` for the internal case, I think this makes the most sense.